src/resolvers/dns.rs

```diff
@@ -587,6 +654,63 @@ mod test {
 
             addr
         }
+
+        async fn run(self) -> SocketAddr {
+            Self::start_server(self.data.build_catalog()).await
+        }
+
+        async fn run_updateable(self) -> UpdateableServer {
+            let catalog =
+                UpdateableCatalog(Arc::new(tokio::sync::Mutex::new(self.data.build_catalog())));
+            let address = Self::start_server(catalog.clone()).await;
+            UpdateableServer {
+                data: self.data,
+                catalog,
+                address,
+            }
+        }
+    }
+
+    #[derive(Clone)]
+    struct UpdateableCatalog(Arc<tokio::sync::Mutex<Catalog>>);
+
+    #[async_trait::async_trait]
+    impl RequestHandler for UpdateableCatalog {
+        async fn handle_request<R: ResponseHandler>(
+            &self,
+            request: &Request,
+            response_handle: R,
+        ) -> ResponseInfo {
+            self.0
+                .lock()
+                .await
+                .handle_request(request, response_handle)
+                .await
+        }
+    }
+
+    struct UpdateableServer {
+        data: DnsServerData,
+        address: SocketAddr,
+        catalog: UpdateableCatalog,
+    }
+
+    impl UpdateableServer {
+        async fn remove_backend(&mut self, name: &str) {
+            self.data.remove_backend(name);
+            self.catalog.0.lock().await.upsert(
+                LowerName::from_str(&self.data.domain).unwrap(),
+                self.data.build_authority(),
+            );
```

Comment:
nit, this could probably be pulled into a little helper and used in both `add_backend` and `remove_backend` since it's identical